### PR TITLE
NPCs use their indoor voices when not in danger

### PR DESCRIPTION
**Pull request by RenechCDDA**
#### Summary
Balance "NPCs don't scream at loudness 16 when just saying normal stuff, unless they're in danger"

#### Purpose of change
Mentioned in #71865 (does not close issue)

NPCs yelling like maniacs in the middle of the night because they wanted to step past you. Then yelling like maniacs to apologize to all the people they just woke up. Brilliant.

#### Describe the solution
Replaced the number `16` with ~~`6` if not in danger~~ math!

Calculate a good volume to use instead of magic numbers, based on whether the player will be able to hear (speak louder if PC is far away)/NPC wants the player to hear. Respects if friends are sleeping (intentionally whispers to avoid waking them up).

#### Describe alternatives you've considered
Why don't we make a slider for the player to micromanage how loud they want their NPCs to yell? (this is not a serious question)

#### Testing
Confirmed it compiles and launches only

#### Additional context
Technically speaking if get_shout_volume() is below 16, they were actually *louder* when speaking normally during danger 🤔

### CODE REVIEW

**Comment by IdleSol** [comment]
If I understand correctly what the tests say. It only tests melee kills ```player.melee_attack```. Correct me if I'm wrong.

But would there be a similar result if a firearm was used?

In the situation in the link. I received guilt if I killed a child up to 5 tiles away and didn't if the distance was more than 6.

I checked in manual mode. In that situation, the death of the NPC gives the guilt. But I checked only the standard relationship of the NPC to the player.

**Comment by Nebnis** [comment]
>Why don't we make a slider for the player to micromanage how loud they want their NPCs to yell? (this is not a serious question)

I mean, i would love this. 

"Something like try to be quiet as much as possible but scream really loud if you in danger and alone(not following player)."

**Comment by RenechCDDA** [comment]
> > Why don't we make a slider for the player to micromanage how loud they want their NPCs to yell? (this is not a serious question)
> 
> I mean, i would love this. Something like:
> 
> "Try to be quiet as much as possible but scream really loud if you in danger and alone(not following player)."

Exactly why it is going to be done programmatically and not by having the player repeatedly change settings for things the NPCs should be smart enough to handle.

**Comment by XygenSS** [comment]
Can this be done with the player character as well?

**Comment by RenechCDDA** [comment]
> Can this be done with the player character as well?

The player does not make verbal sound except when using the `You yell...` command. The player 

As far as I can tell, yelling orders doesn't make any noise either.

So there is nothing to be done here, except maybe *add noise* from talking.

**Comment by XygenSS** [comment]
> As far as I can tell, yelling orders doesn't make any noise either. 
> So there is nothing to be done here, except maybe *add noise* from talking.

Is that really the case? Telling followers to guard or follow via the C menu seems to produce sound equal to your usual yelling volume.

talking to them does seem to be completely silent on your part though, noise definitely should be added.

**Comment by Theawesomeboophis** [comment]
The player talking should definitely make sound.

**Comment by Theawesomeboophis** [comment]
> >Why don't we make a slider for the player to micromanage how loud they want their NPCs to yell? (this is not a serious question)
> 
> I mean, i would love this.  Something like:
> 
> "Try to be quiet as much as possible but scream really loud if you in danger and alone(not following player)."

We should probably have a be quiet command so they're not screaming during night raids and stuff and alerting everything nearby. Because stealthing with npcs is kinda hard when they're yelling at you that they see a zombie 4 miles down the road.

File: src/npc.cpp [hunk 1]
```diff
@@ -1751,11 +1752,52 @@ void npc::say( const std::string &line, const sounds::sound_t spriority ) const
...
+
+    // Don't wake up friends when using our indoor voice
+    for( const auto &bunk_buddy : get_cached_friends() ) {
+        if( !( bunk_buddy.lock() && bunk_buddy.lock()->is_npc() ) ) {
+            continue;
+        }
+        if( !( bunk_buddy.lock() && bunk_buddy.lock()->is_avatar() ) ) {
+            continue;
+        }
+        if( dynamic_cast<const Character &>( *bunk_buddy.lock() ).has_effect( effect_sleep ) ) {
```
Comment (akrieger) [inline src/npc.cpp:1789]:
```suggestion
        Character* char_buddy = nullptr;
        Creature* buddy = bunk_buddy.lock();
        if ( buddy ) {
          char_buddy = dynamic_cast<Character*>( buddy );
        }
        if ( !char_buddy ) {
          continue;
        }
        if( char_buddy->has_effect( effect_sleep ) ) {
        /* etc */
```

**Comment by Montimorency** [comment]
> > Can this be done with the player character as well?
> 
> The player does not make verbal sound except when using the `You yell...` command. (And working out, and casting verbal spells, and occasionally when in pain. But none of those are terribly relevant.)
> 
> As far as I can tell, yelling orders doesn't make any noise either.
> 
> So there is nothing to be done here, except maybe _add noise_ from talking.

Are we sure about that? 

2543r
235gw4
ag433

_"Everyone take it easy", Dumbledore said calmly._

**Comment by RenechCDDA** [comment]
You're right, it looks like the player actually shouts for those commands. I didn't find it because it's in npctalk.cpp, the last place I expected to find the player talking.

Any problems you find there should be made into a new issue though

**Pull request merged**